pkg/syncer/pv_to_vvol_mapping.go

```diff
@@ -0,0 +1,157 @@
+		},
+	}
+
+	querySelection := cnstypes.CnsQuerySelection{
+		Names: []string{
+			string(cnstypes.QuerySelectionNameTypeBackingObjectDetails),
+		},
+	}
+	queryAllResult, err := metadataSyncer.volumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
+	if err != nil {
+		log.Errorf("csiGetPVtoVvolMapping: failed to QueryAllVolume with err=%+v", err.Error())
+		return
+	}
+
+	// Get K8s PVs in State "Bound".
+	k8sPVs, err := getBoundPVs(ctx, metadataSyncer)
+	if err != nil {
+		log.Errorf("csiGetPVtoVvolMapping: Failed to get PVs from kubernetes. Err: %+v", err)
+		return
+	}
+
+	// volumeHandleToPvcMap maps pv.Spec.CSI.VolumeHandle to the pvc object which
+	// bounded to the pv.
+	volumeHandleToPvcMap := make(volumeHandlePVCMap, len(k8sPVs))
+
+	for _, pv := range k8sPVs {
+		if pv.Spec.ClaimRef != nil && pv.Status.Phase == v1.VolumeBound {
+			pvc, err := metadataSyncer.pvcLister.PersistentVolumeClaims(
+				pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
+			if err != nil {
+				log.Warnf("csiGetPVtoVvolMapping: Failed to get pvc for namespace %s and name %s. err=%+v",
+					pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name, err)
+				continue
+			}
+			volumeHandleToPvcMap[pv.Spec.CSI.VolumeHandle] = pvc
+			log.Debugf("csiGetPVtoVvolMapping: pvc %s/%s is backed by pv %s volumeHandle %s",
+				pvc.Namespace, pvc.Name, pv.Name, pv.Spec.CSI.VolumeHandle)
+		}
+	}
+
+	// pv to vvol Map maps vol.VolumeId.Id to backingobjectId.
+	volumeIdToBackingObjectIdMap := make(volumeIdHealthStatusMap, len(queryAllResult.Volumes))
+
+	for _, vol := range queryAllResult.Volumes {
+		volumeIdToBackingObjectIdMap[vol.VolumeId.Id] = vol.BackingObjectDetails.(*cnstypes.CnsBlockBackingDetails).BackingDiskId
+	}
+
+	for volID, pvc := range volumeHandleToPvcMap {
+		pvTovvolPair := volID + ":" + volumeIdToBackingObjectIdMap[volID]
+		updatePVtoVvolMappingStatus(ctx, k8sclient, pvc, pvTovvolPair)
+	}
+
+	log.Info("csiGetPVtoVvolMapping: end")
+}
+
+func parseMappingString(pvTovvolPair string) (string, string, error) {
+	log := logger.GetLogger(ctx)
+	split := strings.Split(pvTovvolPair, ":")
+	if len(split) != 2 {
+		log.Error("Failed to parse mapping string")
+		return "", "", errors.New("Failed to parse mapping string")
+	}
+	return split[0], split[1], nil
+}
+
+func updatePVtoVvolMappingStatus(ctx context.Context, k8sclient clientset.Interface,
+	pvc *v1.PersistentVolumeClaim, pvTovvolPair string) {
+	log := logger.GetLogger(ctx)
+
+	uuid, backingDiskObjectId, err := parseMappingString(pvTovvolPair)
+	if err != nil {
+		log.Errorf("updatePVtoVvolMappingStatus: Failed to parse mapping pair for pvc %s/%s with err:%+v",
+			pvc.Namespace, pvc.Name, err)
+	}
+	val, found := pvc.Annotations[annPVtoVVOL+ "/" + uuid]
+
+	if !found || val != backingDiskObjectId {
+		// VolumeHealth annotation on pvc is changed, set it to new value.
+		metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annPVtoVVOL+ "/" + uuid, backingObjectId)
```

Comment:
Should there be a ":" between `uuid` and `backingObjectId`?

Reply:
We use "cns.vmware.com/pvuuid" as the key, and the annotation is set as:
metav1.SetMetaDataAnnotation(&pvc.ObjectMeta, annPVtoVVOL+ "/" + uuid, backingObjectId). There is no need to add ":"